hw3/heat.cpp

```diff
@@ -0,0 +1,217 @@
+using GMat3 = GpuMatrix3;
+using CMat3 = CpuMatrix3;
+using GMat = GpuMatrix;
+using CMat = CpuMatrix;
+using namespace cimg_library;
+
+__global__ void BlurrKernel(const unsigned char *in, int width, int height, const GMat Filter, unsigned char *out);
+__global__ void L2Kernel(const unsigned char *blurImg, int widht, int height, const unsigned char *origImg, GMat L2);
+__global__ void reduce(GMat dOut, GMat dIn);
+
+
+__global__ void BlurrKernel(const unsigned char *in, int width, int height, const GMat Filter, unsigned char *out)
+{
+   //Image width by height
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   float3 pixelValue;
+   pixelValue.x = 0.0f;
+   pixelValue.y = 0.0f;
+   pixelValue.z = 0.0f;
+   //Edge of image
+   if(row == 0 || row == height-1 || col == 0 || col == width-1)
+   {
+      //Load rgb pixel values for: input[row][col]
+      pixelValue.x = (float)(in[row * width + col]);
+      pixelValue.y = (float)(in[(row + height) * width + col]);
+      pixelValue.z = (float)(in[(row + height*2) * width + col]);
+       
+   }
+   else
+   {
+      for(int u = -1; u < 2; u ++)
+      {
+         for(int v = -1; v < 2; v++)
+         {
+            //Enumerate rgb pixel values for: input[row+u][col+v]*Filter[u+1][v+1]
+            pixelValue.x += (float)(in[(row+u) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+            pixelValue.y += (float)(in[((row+u) + height) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+            pixelValue.z += (float)(in[((row+u) + height*2) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+         }
+       }
+
+   }
+    
+   //Load rgb values into output[row][col]
+   out[row * width + col] = pixelValue.x > 255 ? 255 : pixelValue.x;
+   out[(row + height) * width + col] = pixelValue.y > 255 ? 255 : pixelValue.y;
+   out[(row + height*2) * width + col] = pixelValue.z > 255 ? 255 : pixelValue.z;
+}
+
+
+__global__ void L2Kernel(const unsigned char *blurImg, int width, int height, const unsigned char *origImg, GMat L2)
+{
+    int row = blockIdx.y * blockDim.y + threadIdx.y;
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float3 pixelValueB;
+    pixelValueB.x = 0.0f;
+    pixelValueB.y = 0.0f;
+    pixelValueB.z = 0.0f;
+    pixelValueB.x = (float)(blurImg[row * width + col]);
+    pixelValueB.y = (float)(blurImg[(row + height) * width + col]);
+    pixelValueB.z = (float)(blurImg[(row + height*2) * width + col]);
+
+    float3 pixelValueO;
+    pixelValueO.x = 0.0f;
+    pixelValueO.y = 0.0f;
+    pixelValueO.z = 0.0f;
+    pixelValueO.x = (float)(origImg[row * width + col]);
+    pixelValueO.y = (float)(origImg[(row + height) * width + col]);
+    pixelValueO.z = (float)(origImg[(row + height*2) * width + col]);
+
+    float temp = abs(pixelValueB.x - pixelValueO.x) + abs(pixelValueB.y - pixelValueO.y) + abs(pixelValueB.z - pixelValueO.z);
+    temp = temp/3.f;
+    temp *= temp;
```

Comment:
I'm sorry, this is the L2 norm. 

You need (pixelValueB.x - pixelValue0.x)^2 etc
then you have to sum the temp with reduce and then sqrt.